---
- name: Configure PX-Backup Schedules
  hosts: localhost
  gather_facts: false

  vars_files:
    - "{{ inventory_dir }}/group_vars/common/all.yaml"
    - "{{ inventory_dir }}/group_vars/backup_schedule/delete_vars.yaml"

  pre_tasks:
    - name: Validate required variables
      assert:
        that:
          - px_backup_api_url is defined
          - org_id is defined
          - schedule_deletes is defined
        fail_msg: "Required variables must be defined"
  
  tasks:

    - name: Debug passed variables
      debug:
        msg:
          - "Schedules: {{ schedule_deletes | to_nice_yaml }}"

    - name: Login and fetch Px-Backup token
      include_tasks: "{{ playbook_dir | dirname }}/auth/auth.yaml"

    - name: Delete Backup Schedules
      block:
        - name: Delete Backup Schedule
          backup_schedule:
            operation: DELETE
            api_url: "{{ px_backup_api_url }}"
            token: "{{ px_backup_token }}"
            org_id: "{{ org_id }}"
            name: "{{ item.name | default(omit) }}"
            uid: "{{ item.uid | default(omit) }}"
            backup_object_type: "{{ item.backup_object_type | default(omit) }}"
            exclude_objects: "{{ item.exclude_objects | default(omit) }}"
            include_objects: "{{ item.include_objects | default(omit) }}"
            include_filter: "{{ item.include_filter | default(omit) }}"
            exclude_filter: "{{ item.exclude_filter | default(omit) }}"
            cluster_scope: "{{ item.cluster_scope | default(omit) }}"
          register: delete_result
          loop: "{{ schedule_deletes }}"

        # Display successful deletes
        - name: Display successful deletes
          debug:
            msg: "Successfully deleted backup '{{ item.item }}'"
          loop: "{{ delete_result.results }}"
          when: not item.failed

        # Display failures
        - name: Display failed deletes
          debug:
            msg: "Failed to delete schedule '{{ item.item }}': {{ item.msg }}"
          loop: "{{ delete_result.results }}"
          when: item.failed

      rescue:
        - name: Display error details
          debug:
            msg: "Failed to delete schedules: {{ 'One or more items failed' if delete_result.results is defined else delete_result.msg | default('Unknown error occurred') }}"
          when: delete_result is defined

        - name: Fail with error message
          fail:
            msg: "Failed to delete schedules. See above for details."
